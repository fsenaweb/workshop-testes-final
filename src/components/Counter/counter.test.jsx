import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import user from "@testing-library/user-event";

import Counter from ".";

describe("<Counter />", () => {
  describe("Inicializando com as propriedades preenchidas", () => {
    beforeEach(() => {
      render(<Counter defaultCount={10} description="Título" />);
    });

    it('Deve renderizar o texto "Contagem atual: 10"', () => {
      expect(screen.getByText("Contagem atual: 10")).toBeInTheDocument();
    });

    it('Deve renderizar o texto "Título"', () => {
      expect(screen.getByText(/título/i)).toBeInTheDocument();
    });
  });

  it('Quando o incrementor muda para 5, o texto "Contagem atual: 15" deve ser renderizado', async () => {
    render(<Counter defaultCount={10} description="Título" />);

    const button = screen.getByRole("button", { name: "adicionar número" });
    const input = screen.getByTestId("input");

    fireEvent.change(input, { target: { value: "5" } });
    user.click(button);

    await waitFor(() => {
      expect(screen.getByText("Contagem atual: 15")).toBeInTheDocument();
    });
  });

  it("Deve incrementar e decrementar a contagem corretamente", () => {
    render(<Counter description="Título" defaultCount={0} />);

    const buttonIncrement = screen.getByRole("button", {
      name: "adicionar número",
    });
    const buttonDecrement = screen.getByRole("button", {
      name: "subtrair número",
    });
    const input = screen.getByTestId("input");

    fireEvent.change(input, { target: { value: "2" } });
    fireEvent.click(buttonIncrement);

    expect(screen.getByText("Contagem atual: 2")).toBeInTheDocument();

    fireEvent.click(buttonDecrement);

    expect(screen.getByText("Contagem atual: 0")).toBeInTheDocument();
  });

  it('Quando o incrementor for 0, o texto "Contagem atual: 0" deve ser renderizado', () => {
      render(<Counter defaultCount={0} description="Título" />);

      expect(screen.getByText("Contagem atual: 0")).toBeInTheDocument();
  });

  it("Quando não passar um defaultCount, o texto 'Contagem atual: 0' deve ser renderizado", () => {
    const { container } = render(<Counter description="Título" />);

    expect(screen.getByText("Contagem atual: 0")).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
