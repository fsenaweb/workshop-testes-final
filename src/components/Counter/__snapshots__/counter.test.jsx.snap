// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> Quando não passar um defaultCount, o texto 'Contagem atual: 0' deve ser renderizado 1`] = `
<div>
  <div
    class="text-body-secondary py-3 my-4 border-top"
  >
    <h3
      class="mb-3 px-3"
    >
      Título
    </h3>
    <div
      class="row mx-3 d-flex align-items-center justify-content-evenly "
    >
      <div
        class="col-1 mb-3 px-3"
      >
        <label
          class="form-label"
        >
          Incremento:
          <input
            class="form-control form-control-sm"
            data-testid="input"
            type="text"
            value="1"
          />
        </label>
      </div>
      <div
        class="col-11 d-flex gap-2"
      >
        <button
          aria-label="subtrair número"
          class="btn btn-primary btn-sm"
        >
          -
        </button>
        <button
          aria-label="adicionar número"
          class="btn btn-primary btn-sm ml-3"
        >
          +
        </button>
      </div>
      <h4
        class="mt-1"
      >
         Contagem atual: 
        0
      </h4>
    </div>
  </div>
</div>
`;
